Update frontend to dmd/stable

ibuclaw opened the pull request:
Only got as far as linking, no testing runtime problems yet...

ibuclaw committed:
Update frontend to dmd/stable

ibuclaw commented:
Compiler _seems_ fine now when building the parts of druntime and phobos it can.  No compile-time ICE or failures when building druntime/phobos.  Only reported missing symbols in object.d

ibuclaw committed:
Update to 2.081.0-beta.2

ibuclaw committed:
Update druntime object module to dmd/stable

Geod24 commented:
That's 2.081 right ? 👍

ibuclaw commented:
@Geod24 unless a release happened in the last seven hours since I looked, then it's still 2.081-beta.

ibuclaw commented:
The first patch should be split up between frontend and glue changes. Some parts can be cherry picked into the C++ branch.

ibuclaw committed:
Update calls to frontend from codegen

ibuclaw committed:
Implement vmessage for verbose messages

ibuclaw committed:
Update accessors to the vtable symbol

ibuclaw committed:
Update functions called from the frontend

ibuclaw committed:
Update interface to Global structure in codegen

ibuclaw committed:
Set global showColumns parameter.

ibuclaw committed:
Move unnesting of functions to get_symbol_decl.

ibuclaw committed:
Update druntime bindings to dmd/stable

ibuclaw committed:
Update druntime to dmd/stable

ibuclaw committed:
Update druntime testsuite to dmd/stable

ibuclaw committed:
Update phobos to dmd/stable

ibuclaw committed:
Update testsuite to dmd/stable

ibuclaw committed:
Work around type merging bug in front-end (issue 19021)

ibuclaw committed:
Fix GDC testsuite tests for updated frontend version.

ibuclaw commented:
There's a number of ICEs introduced by the new frontend.  Surprisingly a couple look suspiciously like memory corruption.  Possibly coming from dinterpret.d, where in one example there's a `StructLiteralExp` with pointers to what was then the stack.

```
# of expected passes		30806
# of unexpected failures	296
# of unresolved testcases	2924
```

ibuclaw committed:
Fix GDC for updated frontend version.

ibuclaw committed:
Add options for no RTTI, Exceptions, or D runtime library

ibuclaw committed:
Add aliasing and warning flags to D compile command

ibuclaw committed:
Add support for more CTFE math intrinsics.

ibuclaw committed:
Support COMPILED_IMPORTS in testsuite, remove fix-up for include paths

ibuclaw committed:
Add support for static delegate variables

ibuclaw committed:
Fix C++ dtor mangling.

ibuclaw committed:
Implicitly set C main return type as 'int'

ibuclaw committed:
Implement pragma crt_constructor and drt_constructor

ibuclaw committed:
Update to 2.081.0-rc.1

ibuclaw commented:
This 3 dozen commit mess is the reason why I don't do "release" merges anymore.  Still many bugs to fix, still memory corruption bug in interpret3.d, many implementation regressions pushed upstream though.

```
# of expected passes		31005
# of unexpected failures	44
# of unresolved testcases	2868
```

ibuclaw commented:
Thousands of unresolved tests were due to a link error have seen previously (there's an open upstream bug report on it).

Reverted, and now we're down to something more manageable.
```
# of expected passes		33729
# of unexpected failures	76
# of unresolved testcases	112
```

ibuclaw committed:
Fix handling of tristate commandline option flags.

ibuclaw committed:
Add builtin clonglong and culonglong types

ibuclaw committed:
Add transition options for integer promotions and dip1008

ibuclaw committed:
Add support for -fmain switch

ibuclaw committed:
Don't fill vtbl or write out overriding vtables for interfaces

ibuclaw committed:
Change traceback format to match dmd

ibuclaw committed:
Evaluate side effects in empty struct comparisons

ibuclaw commented:
Looks like just around 4 things left to look at.
- ICE in testsuite caused by memory corruption.
- ICE in building unittests - bad dwarf codegen by the looks.
- Check for new runtime library functions that should be called, add missing tests.
- Resync with upstream, as its been another week since last.

ibuclaw committed:
Fix DWARF ICE declaring nested interface with static members

ibuclaw commented:
ICE in unittests is reproducible on master:
```
void test()
{
  interface I {
    final void foo() { }
  }
}
```
Happens because dwarf2out doesn't handle DW interface types in `prune_unused_types_walk_local_classes()`, added case to the gcc patches, however alternatively we could just not mark D interface types with `RECORD_IS_INTERFACE` in the first place, so its just treated as a regular class according to debug info.

ibuclaw committed:
Use call _d_print_throwable during unwind failure and add message to terminate()

ibuclaw committed:
Move Id structure to separate header

ibuclaw committed:
Update to 2.081.1

ibuclaw commented:
Looks like just the memory corruption to address.

ibuclaw commented:
Got it, `StructLiteralExp` has a self reference field that becomes garbage if you emplace an instance on the stack.  It seems that dmd never ICE's because it doesn't use `sle->origin` in its data visitor pass, but does in its expression visitor pass.

There's probably a way to hit the same bug in dmd, just wouldn't know the code to trigger it (yet).

ibuclaw commented:
Green!

ibuclaw commented:
Rebased against 2.081, plus one patch to fix segfault. https://github.com/dlang/dmd/pull/8483

ibuclaw commented:
Writing out the change logs for each commit and last bits of commit squashing.

belka-ew commented:
2.081.1 is released.

ibuclaw commented:
Yep, that has already been done last night. This patch series was already synced up to the beta, the change to release was just an update of the version file.

ibuclaw commented:
I would also invite you to review commit by commit the parts modified on gdc specific files, I'll write up a list of notable changes.

belka-ew commented:
> I would also invite you to review commit by commit the parts modified on gdc specific files, I'll write up a list of notable changes.

I think the ChangeLog is such a list.

ibuclaw commented:
Some changes may not be so obvious.  Or maybe they are (e.g: https://github.com/D-Programming-GDC/GDC/pull/683/commits/1601965f35e8b0ee2940f762acca336ed20d3415)

ibuclaw commented:
Added changelogs, should do one last sweep of new functions, variables, macros, to ensure they have some documentation.

ibuclaw committed:
Add -frelease when frontend tree checking is disabled

ibuclaw committed:
Remove lowering of array comparison and non-scalar switch statements

ibuclaw committed:
Implement standalone assert contracts

ibuclaw commented:
Sweep done, found a couple more things to add on top.

Removed runtime function definitions that are now lowered by the frontend (`__cmp`, `__switch`).
Adding support for druntime-less assert contracts.

ibuclaw commented:
Full list of changes, and noting of patches that can be cherry picked to C++ branch.

- Frontend and Library is v2.081.1
- Updated calls to frontend functions, moved a few functions from d-frontend.cc to dmd frontend (using `merge2` and `isCPPclass` ok for stable)
- Added vmessage (ok for stable)
- Updated interface for generating vtblsym (could be backported)
- Updated Target interface
- Bug fixes in codegen exposed by new tests (ok for stable)
- Added more CTFE intrinsics (ok for stable)
- Allow static delegates (ok for stable)
- Support more options in testsuite, removed include flag for testsuite source directory, making `EXTRA_FILES` explicitly needed (some parts ok for stable)
- Implemented pragma crt con/de-structor
- Added clonglong and clonglong to builtins module (ok for stable)
- Added/Updated command line options (`-fmain` ok for stable)
- Traceback messages updated to match dmd/druntime format.
- Removed array cmp and switch runtime functions from the codegen.
- Added "betterC" support (could be backported, to see if we can compile the frontend with `-fno-rtti -fno-exceptions`).

@jpf91 / @belka-ew - ok to merge?

ibuclaw committed:
Check type codes match before checking equivalence

ibuclaw committed:
Move dfrontend headers to dmd directory

belka-ew commented:
I think, I found a regression:
```d
module dub.generators.generator;

struct Package
{
    @property string name() const
    {
        return null;
    }
}

void ProjectGenerator()
{
    string[string] targets;

    void collectDependencies(ref string)
    {
    }

    collectDependencies(targets[Package.init.name]);
}
```
gives:
> source/dub/generators/generator.d: In function ‘ProjectGenerator’:
> source/dub/generators/generator.d:19:24: internal compiler error: in add_expr, at tree.c:7454
>     collectDependencies(targets[Package.init.name]);
>                         ^
> 0x6f7308 ???
> 	../sysdeps/x86_64/start.S:118

ibuclaw commented:
@belka-ew - I can reproduce on stable branch, so was not introduced by this change.

Should raise a bug report though.

belka-ew commented:
ok. everything else looks fine

ibuclaw commented:
OK, hitting merge, have plenty of time to fix bugs and iron out other parts.  At least with this in we can do weekly merge updates with dmd. :-)

Though having a quick look at the size/frequency of `master->stable` merges, will probably want to switch to being in sync with master when the next merge happens.

https://github.com/dlang/dmd/commit/4473071207c78d691f7ed2ec4713172663f3bea8 88 days ago
https://github.com/dlang/dmd/commit/b591bdbf378f20d272ab9f212d124c813c1d963c 29 days ago

As the idea was to catch problems introduced by dmd early in the development cycle.
